expr/expr.go

```diff
@@ -24,209 +25,237 @@ var ErrNoSuchFunction = errors.New("no such function")
 var ErrNotContainer = errors.New("cannot apply in to a non-container")
 var ErrBadCast = errors.New("bad cast")
 
-type NativeEvaluator func(*zng.Record) (zngnative.Value, error)
-
-// CompileExpr tries to compile the given Expression into a function
-// that evalutes the expression against a provided Record.  Returns an
+// CompileExpr compiles the given Expression into an object
+// that evalutes the expression against a provided Record.  It returns an
 // error if compilation fails for any reason.
 //
-// This is currently not particularly optimized -- it creates a bunch
-// of closures and every evaluation involves some allocations.
-// Eventually, we could optimize this by compiling a particular
-// Expression for a particular TypeRecord into a series of byte codes
-// that could be implemented by a simple stack-based evaluator much
-// more efficiently.  ZNG unions are a challenge for this approach, but
-// we could fail back to the "slow path" implemented here if an
-// expression ever touches a union.
-func CompileExpr(node ast.Expression) (ExpressionEvaluator, error) {
-	ne, err := compileNative(node)
-	if err != nil {
-		return nil, err
-	}
-
-	return func(rec *zng.Record) (zng.Value, error) {
-		nv, err := ne(rec)
-		if err != nil {
-			return zng.Value{}, err
-		}
-
-		return nv.ToZngValue()
-	}, nil
+// This is the "intepreted slow path" of the analytics engine.
+// It is completely adaptive to the dynamic type system of zng so there
+// are a lot of types checks and run-tme decisions made according to type.
```

Comment:
```suggestion
// This is the "interpreted slow path" of the analytics engine.
// It adapts to the dynamic type system of zng, so many
// checks and run-time decisions are made according to type.
```

Reply:
This changes the meaning a bit but I'll take a stab at a rewrite of the confusing sentence...